component/prometheus/fanout.go

```diff
@@ -0,0 +1,131 @@
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/metadata"
+
+	"github.com/prometheus/prometheus/storage"
+)
+
+var _ storage.Appendable = (*Fanout)(nil)
+
+type intercept func(ref storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, labels.Labels, int64, float64, error)
+
+// Fanout supports the default Flow style of appendables since it can go to multiple outputs. It also allows the intercepting of appends.
+type Fanout struct {
+	mut sync.RWMutex
+	// intercept allows one to intercept the series before it fans out to make any changes. If labels.Labels returns nil the series is not propagated.
+	// Intercept shouuld be thread safe and can be called across appenders.
+	intercept func(ref storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, labels.Labels, int64, float64, error)
+
```

Comment:
Have you considered making a separate storage.Appendable interface just for intercept? It would behave similarly to the middleware pattern at that point. As it is now, it feels like Fanout is responsible for a little too much.

Reply:
I originally did have more specific appendables, but it's really handy to reuse the fanout for tests. There was a fair amount of duplicated code so I refactored to use one that allows some middleware in between.

Reply:
I'd like a second opinion from @tpaschalis here. My gut feeling is that we shouldn't put both intercept + fanout logic in the same thing, but I don't know if I'm being too idealistic with this.

Reply:
I think I'd be in favor of a separate appendable here.

If it's only the testing code that gets more complicated by splitting the intercept+fanout logic, then maybe we should give it a try, since we won't be altering the test code that much (it's already tested upstream and we only need to validate our component's behavior).

Reply:
This is resolved now.